### PR TITLE
Add messageEdited and messagedDeleted events to ChatAdapter. Update the unread message badge when a message is deleted.

### PR DESCRIPTION
**Pull request by JamesBurnside**
# What

- Add messageEdited and messagedDeleted events to the ChatAdapter.
- Use these events to update the unread message badge in the callwithchat composite when a message is deleted.

# Why

Unread messages badge should update when an unread chat message is deleted.

# How Tested

Verified locally deleting a chat messages decrements the unread message count to other participants in the callwithchat

### CODE REVIEW

File: packages/communication-react/review/beta/communication-react.api.md [hunk 1]
```diff
@@ -2542,6 +2554,9 @@ export const DEFAULT_COMPOSITE_ICONS: {
     SurveyStarIconFilled: React_2.JSX.Element;
 };
 
+// @public (undocumented)
+export type DeletedChatMessage = Omit<ChatMessage_2, 'message' | 'metadata'>;
```
Comment (JamesBurnside) [inline packages/communication-react/review/beta/communication-react.api.md:2558]: Unsure if this is the best payload to transmit for the event, but it creates the closest type to the other events

**Comment by JamesBurnside** [comment]
Added breaking-change label as the breaking change is adding new string literals to the `on` function which is acceptable level

File: change-beta/@azure-communication-react-0bde713d-dea9-4c4f-818d-5bcde6286354.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "type": "patch",
```
[the remark this answers is not in the document]
Comment (JamesBurnside) [reply change-beta/@azure-communication-react-0bde713d-dea9-4c4f-818d-5bcde6286354.json:2]: Really it should be two PRs, so added two changelog entries; one for the new ChatAdapter methods, and one for the fix to updating the ChatButton unread messages badge

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts [hunk 3]
```diff
@@ -476,6 +509,18 @@ const convertEventToChatMessage = (event: ChatMessageReceivedEvent): ChatMessage
...
 
+const convertEventToDeletedChatMessage = (event: ChatMessageDeletedEvent): DeletedChatMessage => {
+  return {
+    id: event.id,
+    version: event.version,
+    type: convertEventType(event.type),
+    sender: event.sender,
+    senderDisplayName: event.senderDisplayName,
+    sequenceId: '',
+    createdOn: new Date(event.createdOn)
```
Comment (vhuseinova-msft) [inline packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:520]: Do we need to set "deletedOn" here too?

File: packages/communication-react/review/stable/communication-react.api.md [hunk 4]
```diff
@@ -849,6 +853,10 @@ export interface CallWithChatAdapterSubscriptions {
     // (undocumented)
     on(event: 'messageReceived', listener: MessageReceivedListener): void;
     // (undocumented)
+    on(event: 'messageEdited', listener: MessageEditedListener): void;
```
Comment (vhuseinova-msft) [inline packages/communication-react/review/stable/communication-react.api.md:856]: It looks like for the new functionality only `messageDeleted` and `messsageReceived` events are used. Is `messageEdited` event needed?

Comment (JamesBurnside) [reply packages/communication-react/review/stable/communication-react.api.md:856]: It's not used by the chat badge button, but for developer observability it seems like an event we've just missed in exposing -- can remove if we don't want to expose it though

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts (hunk 3 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:520]: 100%. And this means we should remove the `deletedOn: Date` in the MessageDeletedListener as well as its redundant.
I should do the same for editedOn

File: packages/communication-react/review/beta/communication-react.api.md (hunk 1 above)
Comment (AmyL219) [reply packages/communication-react/review/beta/communication-react.api.md:2558]: Is the `undocumented` in the comment `// @public (undocumented)` expected?  
Where does the `DeletedChatMessage` being defined?

File: packages/communication-react/review/stable/communication-react.api.md (hunk 4 above)
Comment (vhuseinova-msft) [reply packages/communication-react/review/stable/communication-react.api.md:856]: `It's not used by the chat badge button, but for developer observability it seems like an event we've just missed in exposing`
True! I'm okay to leave it but more feedback from the team may be needed

File: packages/communication-react/review/beta/communication-react.api.md (hunk 1 above)
Comment (JamesBurnside) [reply packages/communication-react/review/beta/communication-react.api.md:2558]: Removing this, not needed

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts (hunk 3 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts]: Thanks updated! This also meant I could simplify a bunch of pieces and remove the DeletedChatMessage interface as it was redundant also

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts [hunk 5]
```diff
@@ -496,29 +495,33 @@ export class AzureCommunicationChatAdapter implements ChatAdapter {
...
 
-const convertEventToChatMessage = (event: ChatMessageReceivedEvent): ChatMessage => {
+const convertEventToChatMessage = (
+  event: ChatMessageReceivedEvent | ChatMessageEditedEvent | ChatMessageDeletedEvent
+): ChatMessage => {
   return {
     id: event.id,
     version: event.version,
-    content: { message: event.message },
+    content: isChatMessageDeletedEvent(event) ? undefined : { message: event.message },
```
Comment (vhuseinova-msft) [inline packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:504]: I'm wondering if Chat DSK returns content for `messageDeleted` event and if so, if we need to be consistent here

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts [hunk 6]
```diff
@@ -496,29 +495,33 @@ export class AzureCommunicationChatAdapter implements ChatAdapter {
...
     type: convertEventType(event.type),
     sender: event.sender,
     senderDisplayName: event.senderDisplayName,
     sequenceId: '',
-    createdOn: new Date(event.createdOn)
+    createdOn: new Date(event.createdOn),
+    editedOn: isChatMessageEditedEvent(event) ? event.editedOn : undefined,
```
Comment (vhuseinova-msft) [inline packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:510]: This property can be returned for `messageDeleted` event from ChatSDK and we probably would want to be consistent

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts [hunk 7]
```diff
@@ -388,6 +392,26 @@ export class AzureCommunicationChatAdapter implements ChatAdapter {
...
   }
 
+  private messageEditedListener(event: ChatMessageEditedEvent): void {
+    const isCurrentChatAdapterThread = event.threadId === this.chatThreadClient.threadId;
+    if (!isCurrentChatAdapterThread) {
+      return;
+    }
+
+    const message = convertEventToChatMessage(event);
+    this.emitter.emit('messageEdited', { message, editedOn: event.editedOn });
```
Comment (vhuseinova-msft) [inline packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:402]: Do we need `editedOn: event.editedOn` here?

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts (hunk 5 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:504]: It explicitly doesn't have it in the payload so had to add this line here

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts (hunk 6 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.ts:510]: It's not in the chatClient event that we are forwarding here, only editedEvent has the editedOn:
image

**Review by vhuseinova-msft — APPROVED** [review]

**Review by dmceachernmsft — APPROVED** [review]

**Pull request merged**